File: frontend/src/v2/api/apps.ts
// frontend/src/v2/api/apps.ts
//
// 应用市场 & 应用域 API 层。
// 后端契约：app/interfaces/api/v1/apps.py
// 所有调用均通过 apiClient，禁止页面层直接调 axios。

import { apiClient } from './client'

// ============================================================================
// 类型定义（与后端 wire 格式保持 snake_case）
// ============================================================================

export interface App {
  id: number
  code: string
  name: string
  category: string
  description: string | null
  config_schema: Record<string, unknown> | null
  icon: string | null
  author: string | null
  version: string | null
  enabled: boolean
  created_at: string | null
  updated_at: string | null
  // include_stats=true 时以下字段非 null
  instance_count: number | null
  active_instance_count: number | null
  total_execution_count: number | null

  // P22: 应用整体健康状态（后端可选返回，缺省按 'unknown' 处理）
  health?: HealthStatus | null

  // drop-frontend: backend has no design for App.rating — see plan §3.4
  // drop-frontend: backend has no design for App.installs — see plan §3.4
  // drop-frontend: backend has no design for App.capabilities — see plan §3.4
}

export type HealthStatus = 'healthy' | 'degraded' | 'unhealthy' | 'down' | 'unknown'

export interface AppListParams {
  category?: string
  enabled_only?: boolean
  include_stats?: boolean
}

export interface AppConfigValidateResult {
  is_valid: boolean
  errors: string[]
}

export interface AppCategoryOption {
  value: string
  label: string
  app_count: number | null
}

type AppCategoryWire =
  | string
  | {
      category?: unknown
      display_name?: unknown
      app_count?: unknown
    }

function normalizeCategoryOption(item: AppCategoryWire): AppCategoryOption | null {
  if (typeof item === 'string') {
    const value = item.trim()
    return value ? { value, label: value, app_count: null } : null
  }

  if (!item || typeof item !== 'object') return null
  const value = typeof item.category === 'string' ? item.category.trim() : ''
  if (!value) return null
  const label =
    typeof item.display_name === 'string' && item.display_name.trim()
      ? item.display_name.trim()
      : value
  const appCount = typeof item.app_count === 'number' ? item.app_count : null
  return { value, label, app_count: appCount }
}

// ============================================================================
// 应用市场接口
// ============================================================================

export async function listApps(params: AppListParams = {}): Promise<App[]> {
  const res = await apiClient.get<{ data: App[] }>('/apps', { params })
  return res.data.data
}

export async function getApp(code: string): Promise<App> {
  const res = await apiClient.get<{ data: App }>(`/apps/${code}`)
  return res.data.data
}

export async function getAppConfigSchema(
  code: string,
): Promise<Record<string, unknown>> {
  const res = await apiClient.get<{ data: Record<string, unknown> }>(
    `/apps/${code}/config-schema`,
  )
  return res.data.data
}

export async function listAppCategories(): Promise<AppCategoryOption[]> {
  const res = await apiClient.get<{ data: AppCategoryWire[] }>('/apps/categories')
  return (res.data.data ?? [])
    .map(normalizeCategoryOption)
    .filter((item): item is AppCategoryOption => item != null)
}

export async function validateAppConfig(
  code: string,
  config: Record<string, unknown>,
): Promise<AppConfigValidateResult> {
  const res = await apiClient.post<{ data: AppConfigValidateResult }>(
    `/apps/${code}/validate`,
    { config },
  )
  return res.data.data
}

// ── 启用 / 停用（P20）────────────────────────────────────────────────────────

export async function enableApp(code: string): Promise<App> {
  const res = await apiClient.post<{ data: App }>(`/apps/${code}/enable`)
  return res.data.data
}

export async function disableApp(code: string): Promise<App> {
  const res = await apiClient.post<{ data: App }>(`/apps/${code}/disable`)
  return res.data.data
}

File: app/interfaces/api/v1/apps.py
"""
应用市场 API

提供应用定义的查询接口
"""
from flask import Blueprint, request, g

from app.di.container import get_container
from app.interfaces.api.middleware.auth import require_auth
from app.shared.response import success, not_found, server_error
from app.shared.utils.logger import get_logger

logger = get_logger(__name__)

# 创建 Blueprint
bp = Blueprint('apps_api_v1', __name__, url_prefix='/api/v1/apps')


def _get_service():
    """从 DI 容器获取 AppDefinitionService"""
    return get_container().app_definition_service()


@bp.route('', methods=['GET'])
@require_auth
def get_apps():
    """获取应用列表"""
    category = request.args.get('category')
    enabled_only = request.args.get('enabled_only', 'true').lower() == 'true'
    include_stats = request.args.get('include_stats', 'false').lower() == 'true'

    service = _get_service()
    apps = service.get_all_apps(
        category=category,
        enabled_only=enabled_only,
        include_stats=include_stats
    )
    return success(data=apps)


@bp.route('/<string:code>', methods=['GET'])
@require_auth
def get_app(code: str):
    """获取应用详情"""
    service = _get_service()
    app = service.get_app_by_code(code)

    if not app:
        return not_found(message=f"应用 {code} 不存在")

    return success(data=app)


@bp.route('/<string:code>/config-schema', methods=['GET'])
@require_auth
def get_config_schema(code: str):
    """获取应用配置表单 JSON Schema"""
    service = _get_service()
    schema = service.get_config_schema(code)

    if not schema:
        return not_found(message=f"应用 {code} 不存在或未定义配置模板")

    return success(data=schema)


@bp.route('/categories', methods=['GET'])
@require_auth
def get_categories():
    """获取所有应用分类"""
    service = _get_service()
    categories = service.get_categories()
    return success(data=categories)


@bp.route('/<string:code>/validate', methods=['POST'])
@require_auth
def validate_config(code: str):
    """验证应用配置"""
    data = request.get_json()
    config = data.get('config', {})

    service = _get_service()
    is_valid, errors = service.validate_app_config(code, config)

    return success(data={
        'is_valid': is_valid,
        'errors': errors
    })


@bp.route('/<string:code>/enable', methods=['POST'])
@require_auth
def enable_app(code: str):
    """启用应用"""
    service = _get_service()
    app = service.set_enabled(code, True)
    if not app:
        return not_found(message=f"应用 {code} 不存在")
    return success(data=app, message="应用已启用")


@bp.route('/<string:code>/disable', methods=['POST'])
@require_auth
def disable_app(code: str):
    """停用应用"""
    service = _get_service()
    app = service.set_enabled(code, False)
    if not app:
        return not_found(message=f"应用 {code} 不存在")
    return success(data=app, message="应用已停用")
